### PR TITLE
choose() takes a bool, chooseAmong() takes a size_t; also fix a numbe…

### PR DESCRIPTION
**Pull request by andralex**
…r of bugs in the original

This is a continuation of https://github.com/D-Programming-Language/phobos/pull/3269 per a discussion with @WalterBright. Instead of defining choose with size_t and any number of ranges, this defines choose for the frequent case (a la `Select` etc) and defines chooseAmong for arbitrary number of ranges, defined in terms of choose. Implementation comes 30% shorter.

There are a few bugfixes, particularly in the copy semantics.

### CODE REVIEW

**Comment by MetaLang** [comment]
Is it worth making the boolean parameter lazy? I've only really ever seen it used for enforce, and it kills inlining, but it may be worth considering. Also, how about an overload that takes a predicate function for dynamic choosing?

File: std/range/package.d [hunk 1]
```diff
@@ -1228,312 +1228,209 @@ unittest
...
+                auto result = this;
+                if (condition) result.r1 = result.r1[begin .. end];
+                else result.r2 = result.r2[begin .. end];
+                return result;
             }
+    }
+    Result result;
+    result.condition = condition;
+    import std.conv : emplace;
+    if (condition) emplace(&result.r1(), r1);
```
Comment (MetaLang) [inline std/range/package.d]: Is there any reason not to move this emplace into `Result` itself?

Comment (andralex) [inline std/range/package.d]: makes sense thx

**Comment by andralex** [comment]
`condition` can't be made lazy without significant cost (both ranges must be copied into `this`) for an unlikely use pattern

**Comment by andralex** [comment]
acted on feedback

**Comment by WalterBright** [comment]
How's the unit test coverage? I.e. at 100%? Otherwise, LGTM.

File: std/range/package.d [hunk 2]
```diff
@@ -1228,312 +1228,212 @@ unittest
...
-                }
-        }
-        return Result(index, rs);
-    }
+Returns:
+    The indexed range. If rs consists of only one range, the return type is an
+    alias of that range's type.
+ */
+auto chooseAmong(Ranges...)(size_t index, Ranges rs)
+if (Ranges.length > 2
```
Comment (MetaLang) [inline std/range/package.d]: Why two separate functions instead of some internal `static if`s?

Comment (andralex) [inline std/range/package.d]: makes the constraint quite a bit easier to understand

**Comment by andralex** [comment]
@WalterBright coverage is 100%

**Pull request merged**